[DEFECT] Folder rename fails with database constraint violation when folder contains many items

### Problem Statement

Folder rename operation fails with a database constraint violation error when the folder contains a large number of contentlets (more than 20 K contentlets). The users are blocked from renaming folders with a large number of items so cannot reorganize the folder names as required.

### Steps to Reproduce

1. Create a folder with 20000+ content items
2. Navigate to Site Browser
3. Right-click folder and choose 'Edit Folder' option
4. Enter the new name and then click Save
5. The operation fails and an error message is shown in the log
```
[13/02/26 21:51:26:835 UTC] ERROR business.FolderAPIImpl: Error renaming folder '/outreach/' with id: 18aea225-cc58-415e-a386-5fbd67d18b13 to name: outreach-v2. Error: ERROR: Cannot delete as this path has children
  Where: PL/pgSQL function check_child_assets() line 13 at RAISE
SQL statement "DELETE from identifier where id = OLD.identifier"
PL/pgSQL function folder_identifier_check() line 8 at SQL statement{
  "SQL": ["delete from folder where folder.inode = ? "],
  "maxRows": [-1],
  "offest": [0],
  "params": "18aea225-cc58-415e-a386-5fbd67d18b13"
}
[13/02/26 21:51:26:835 UTC] ERROR business.FolderAPIImpl: Error renaming folder '/outreach/' with id: 18aea225-cc58-415e-a386-5fbd67d18b13 to name: outreach-v2. Error: ERROR: Cannot delete as this path has children
  Where: PL/pgSQL function check_child_assets() line 13 at RAISE
SQL statement "DELETE from identifier where id = OLD.identifier"
PL/pgSQL function folder_identifier_check() line 8 at SQL statement{
  "SQL": ["delete from folder where folder.inode = ? "],
  "maxRows": [-1],
  "offest": [0],
  "params": "18aea225-cc58-415e-a386-5fbd67d18b13"
}
[13/02/26 21:51:34:793 UTC] ERROR action.EditFolderAction: ERROR: Cannot save folder '/outreach-v2'
com.dotmarketing.exception.DotDataException: ERROR: Cannot delete as this path has children
  Where: PL/pgSQL function check_child_assets() line 13 at RAISE
SQL statement "DELETE from identifier where id = OLD.identifier"
PL/pgSQL function folder_identifier_check() line 8 at SQL statement{
  "SQL": ["delete from folder where folder.inode = ? "],
  "maxRows": [-1],
  "offest": [0],
  "params": "18aea225-cc58-415e-a386-5fbd67d18b13"
}
	at com.dotmarketing.portlets.folders.business.FolderAPIImpl.renameFolder(FolderAPIImpl.java:184) ~[?:?]
	at com.dotmarketing.portlets.folders.action.EditFolderAction._updateFolder(EditFolderAction.java:326) ~[?:?]
	at com.dotmarketing.portlets.folders.action.EditFolderAction.processAction(EditFolderAction.java:121) ~[?:?]
	at com.liferay.portal.struts.PortletRequestProcessor.process(PortletRequestProcessor.java:126) ~[?:?]
	at com.liferay.portlet.StrutsPortlet.processAction(StrutsPortlet.java:120) ~[?:?]
	at com.liferay.portlet.ConcretePortletWrapper.processAction(ConcretePortletWrapper.java:97) ~[?:?]
	at com.liferay.portal.action.LayoutAction._processPortletRequest(LayoutAction.java:194) ~[?:?]
	at com.liferay.portal.action.LayoutAction._processActionRequest(LayoutAction.java:206) ~[?:?]
	at com.liferay.portal.action.LayoutAction.execute(LayoutAction.java:82) ~[?:?]
	at com.dotcms.repackage.org.apache.struts.action.RequestProcessor.processActionPerform(RequestProcessor.java:431) ~[?:?]
	at com.dotcms.repackage.org.apache.struts.action.RequestProcessor.process(RequestProcessor.java:237) ~[?:?]
	at com.liferay.portal.struts.PortalRequestProcessor.process(PortalRequestProcessor.java:149) ~[?:?]
	at com.dotcms.repackage.org.apache.struts.action.ActionServlet.process(ActionServlet.java:1202) ~[?:?]
	at com.dotcms.repackage.org.apache.struts.action.ActionServlet.doPost(ActionServlet.java:430) ~[?:?]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:555) ~[?:?]
	at com.liferay.portal.servlet.MainServlet.callParentService(MainServlet.java:227) ~[?:?]
	at com.liferay.portal.servlet.MainServlet.service(MainServlet.java:427) ~[?:?]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:623) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:199) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.repackage.com.liferay.filters.secure.SecureFilter.doFilter(SecureFilter.java:135) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotmarketing.filters.CMSFilter.doFilterInternal(CMSFilter.java:212) ~[?:?]
	at com.dotmarketing.filters.CMSFilter.doFilter(CMSFilter.java:62) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.filters.interceptor.AbstractWebInterceptorSupportFilter.doFilter(AbstractWebInterceptorSupportFilter.java:90) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.filters.interceptor.AbstractWebInterceptorSupportFilter.doFilter(AbstractWebInterceptorSupportFilter.java:90) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.visitor.filter.servlet.VisitorFilter.doFilter(VisitorFilter.java:81) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.vanityurl.filters.VanityURLFilter.doFilter(VanityURLFilter.java:107) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at org.tuckey.web.filters.urlrewrite.RuleChain.handleRewrite(RuleChain.java:176) ~[?:?]
	at org.tuckey.web.filters.urlrewrite.RuleChain.doRules(RuleChain.java:145) ~[?:?]
	at org.tuckey.web.filters.urlrewrite.UrlRewriter.processRequest(UrlRewriter.java:92) ~[?:?]
	at org.tuckey.web.filters.urlrewrite.UrlRewriteFilter.doFilter(UrlRewriteFilter.java:389) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotmarketing.filters.TimeMachineFilter.doFilter(TimeMachineFilter.java:149) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.filters.interceptor.AbstractWebInterceptorSupportFilter.doFilter(AbstractWebInterceptorSupportFilter.java:90) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotmarketing.filters.ThreadNameFilter.doFilter(ThreadNameFilter.java:88) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotmarketing.filters.CharsetEncodingFilter.doFilter(CharsetEncodingFilter.java:99) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotmarketing.filters.CookiesFilter.doFilter(CookiesFilter.java:53) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:129) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.filters.RequestTrackingFilter.doFilter(RequestTrackingFilter.java:77) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.cost.RequestCostFilter.doFilter(RequestCostFilter.java:65) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at com.dotcms.filters.NormalizationFilter.doFilter(NormalizationFilter.java:89) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144) ~[?:?]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:168) ~[?:?]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[?:?]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482) ~[?:?]
	at com.dotcms.tomcat.redissessions.RedisSessionHandlerValve.invoke(RedisSessionHandlerValve.java:38) ~[?:?]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:130) ~[?:?]
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:656) ~[?:?]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[?:?]
	at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:762) ~[?:?]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[?:?]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:346) ~[?:?]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:397) ~[?:?]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[?:?]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:935) ~[?:?]
	at org.apache.tomcat.util.net.Nio2Endpoint$SocketProcessor.doRun(Nio2Endpoint.java:1730) ~[?:?]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[?:?]
	at org.apache.tomcat.util.net.AbstractEndpoint.processSocket(AbstractEndpoint.java:1426) ~[?:?]
	at org.apache.tomcat.util.net.Nio2Endpoint.setSocketOptions(Nio2Endpoint.java:339) ~[?:?]
	at org.apache.tomcat.util.net.Nio2Endpoint$Nio2Acceptor.completed(Nio2Endpoint.java:477) ~[?:?]
	at org.apache.tomcat.util.net.Nio2Endpoint$Nio2Acceptor.completed(Nio2Endpoint.java:415) ~[?:?]
	at java.base/sun.nio.ch.Invoker.invokeUnchecked(Invoker.java:129) ~[?:?]
	at java.base/sun.nio.ch.Invoker$2.run(Invoker.java:221) ~[?:?]
	at java.base/sun.nio.ch.AsynchronousChannelGroupImpl$1.run(AsynchronousChannelGroupImpl.java:113) ~[?:?]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1189) ~[?:?]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:658) ~[?:?]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: com.dotmarketing.exception.DotDataException: ERROR: Cannot delete as this path has children
  Where: PL/pgSQL function check_child_assets() line 13 at RAISE
SQL statement "DELETE from identifier where id = OLD.identifier"
PL/pgSQL function folder_identifier_check() line 8 at SQL statement{
  "SQL": ["delete from folder where folder.inode = ? "],
  "maxRows": [-1],
  "offest": [0],
  "params": "18aea225-cc58-415e-a386-5fbd67d18b13"
}
	at com.dotmarketing.common.db.DotConnect.loadResult(DotConnect.java:336) ~[?:?]
	at com.dotmarketing.portlets.folders.business.FolderFactoryImpl.delete(FolderFactoryImpl.java:101) ~[?:?]
	at com.dotmarketing.portlets.folders.business.FolderFactoryImpl.renameFolder(FolderFactoryImpl.java:820) ~[?:?]
	at com.dotmarketing.portlets.folders.business.FolderAPIImpl.renameFolder(FolderAPIImpl.java:170) ~[?:?]
	... 100 more
Caused by: org.postgresql.util.PSQLException: ERROR: Cannot delete as this path has children
  Where: PL/pgSQL function check_child_assets() line 13 at RAISE
SQL statement "DELETE from identifier where id = OLD.identifier"
PL/pgSQL function folder_identifier_check() line 8 at SQL statement
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2725) ~[?:?]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2412) ~[?:?]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371) ~[?:?]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:502) ~[?:?]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:419) ~[?:?]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:194) ~[?:?]
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:180) ~[?:?]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44) ~[?:?]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.executeQuery(DotConnect.java:703) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.executeQuery(DotConnect.java:626) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.loadResult(DotConnect.java:334) ~[?:?]
	at com.dotmarketing.portlets.folders.business.FolderFactoryImpl.delete(FolderFactoryImpl.java:101) ~[?:?]
	at com.dotmarketing.portlets.folders.business.FolderFactoryImpl.renameFolder(FolderFactoryImpl.java:820) ~[?:?]
	at com.dotmarketing.portlets.folders.business.FolderAPIImpl.renameFolder(FolderAPIImpl.java:170) ~[?:?]
	... 100 more
```

### Acceptance Criteria

- Folder rename should succeed regardless of the number of items contained, completing as an atomic operation.
- All child identifiers should correctly updated to new parent path.


### dotCMS Version

Evergreen (26.02.11-01)

### Severity

High - Major functionality broken

### Links

Freshdesk ticket [34969](https://helpdesk.dotcms.com/a/tickets/34969)

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java
@@ -677,7 +677,10 @@ public class FolderAPIImpl implements FolderAPI  {
 		final boolean isNew = folder.getInode() == null;
 		//if the folder was renamed, we will need to create a new identifier
 		if (!folder.getName().equalsIgnoreCase(existingID.getAssetName())){
-			folderFactory.renameFolder(folder, folder.getName(), user, respectFrontEndPermissions);
+			if (!folderFactory.renameFolder(folder, folder.getName(), user, respectFrontEndPermissions)) {
+				throw new DotDataException("Could not rename folder '" + existingID.getAssetName()
+						+ "' to '" + folder.getName() + "': a folder with that name already exists.");
+			}
 			// Queue async ES reindex so content under the renamed folder is re-indexed at the new
 			// path. folderFactory.renameFolder() mutates folder.setName(newName), so
 			// folder.getPath() already returns the new path when this runs.

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java
@@ -174,10 +174,11 @@ public class FolderAPIImpl implements FolderAPI  {
 			renamed = folderFactory.renameFolder(folder, newName, user, respectFrontEndPermissions);
 
 			// Nav cache eviction for the folder and sub-tree is handled inside the factory.
-			// NOTE: the factory mutates the passed-in folder via folder.setName(newName), so
-			// folder.getPath() returns the new path here. refreshContentUnderFolder depends on
-			// this side-effect to queue the reindex against the renamed path. Do not refactor
-			// the factory to work on a defensive copy without updating this call site.
+			// NOTE: the factory mutates the passed-in folder: setName(newName), setInode(), and
+			// setIdentifier() are all updated to reflect the newly created folder record.
+			// refreshContentUnderFolder depends on these side-effects to target the renamed path
+			// and the correct new inode/identifier. Do not refactor the factory to work on a
+			// defensive copy without updating this call site.
 			//
 			// Queue async ES reindex. DotReindexStateException is caught here so a transient
 			// reindex-queue failure does not roll back an otherwise successful rename.

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
@@ -8,6 +8,7 @@ import static com.dotmarketing.portlets.folders.business.FolderFactorySql.GET_CO
 import static com.dotmarketing.portlets.folders.business.FolderFactorySql.GET_CONTENT_TYPE_COUNT;
 
 import com.dotcms.browser.BrowserQuery;
+import com.dotcms.business.WrapInTransaction;
 import com.dotcms.system.SimpleMapAppContext;
 import com.dotcms.util.transform.DBTransformer;
 import com.dotcms.util.transform.TransformerLocator;
@@ -52,7 +53,6 @@ import com.liferay.util.StringPool;
 import io.vavr.control.Try;
 import java.io.IOException;
 import java.sql.Connection;
-import java.sql.SQLException;
 import java.sql.Timestamp;
 import java.util.ArrayList;
 import java.util.Collections;
@@ -632,21 +632,14 @@ public class FolderFactoryImpl extends FolderFactory {
     return successOperation.getValue();
   }
 
-  private void updateOtherFolderReferences(final String newFolderInode, final String oldFolderInode) {
-    final DotConnect dotConnect = new DotConnect();
-    try {
-      dotConnect.executeStatement("update structure set folder = '" + newFolderInode
-          + "' where folder = '" + oldFolderInode + "'");
-      dotConnect.executeStatement("update permission set inode_id = '" + newFolderInode
-          + "' where inode_id = '" + oldFolderInode + "'");
-      dotConnect.executeStatement("update permission_reference set asset_id = '"
-          + newFolderInode + "' where asset_id = '" + oldFolderInode + "'");
-
-
-    } catch (SQLException e) {
-      Logger.error(FolderFactoryImpl.class, e.getMessage(), e);
-      throw new DotRuntimeException(e.getMessage(), e);
-    }
+  private void updateOtherFolderReferences(final String newFolderInode, final String oldFolderInode)
+      throws DotDataException {
+    new DotConnect().executeUpdate(
+        "UPDATE structure SET folder = ? WHERE folder = ?", newFolderInode, oldFolderInode);
+    new DotConnect().executeUpdate(
+        "UPDATE permission SET inode_id = ? WHERE inode_id = ?", newFolderInode, oldFolderInode);
+    new DotConnect().executeUpdate(
+        "UPDATE permission_reference SET asset_id = ? WHERE asset_id = ?", newFolderInode, oldFolderInode);
   }
 
   /**
@@ -778,6 +771,7 @@ public class FolderFactoryImpl extends FolderFactory {
 		}
   }
 
+  @WrapInTransaction
   @Override
   protected boolean renameFolder(final Folder folder, final String newName, final User user,
       final boolean respectFrontEndPermissions) throws DotDataException, DotSecurityException {
@@ -816,7 +810,7 @@ public class FolderFactoryImpl extends FolderFactory {
     // based on assetType:hostname:parentPath:folderName — changing the URL changes the identity.
     final Folder newFolder;
     try {
-      newFolder = getNewFolderRecord(folder, APILocator.systemUser(), parentPath, hostId);
+      newFolder = getNewFolderRecord(folder, user, parentPath, hostId);
     } catch (final DotDataException | RuntimeException e) {
       folder.setName(ident.getAssetName()); // restore original name
       // A concurrent rename to the same destination path can race past the collision check and
@@ -864,6 +858,12 @@ public class FolderFactoryImpl extends FolderFactory {
     // folder.getName() was already set to newName above.
     folder.setInode(newFolder.getInode());
     folder.setIdentifier(newFolder.getIdentifier());
+
+    // Evict the new folder's own identifier cache entry. clearIdentifierCacheForSubtree()
+    // only covers items whose parent_path starts with newPath (the folder's children), not the
+    // folder itself (whose parent_path is parentPath). Without this eviction, a prior URI-keyed
+    // entry for the same path could be served until natural LRU expiry.
+    CacheLocator.getIdentifierCache().removeFromCacheByIdentifier(newFolder.getIdentifier());
 
     return true;
   }

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
@@ -805,41 +805,20 @@ public class FolderFactoryImpl extends FolderFactory {
     // Snapshot sub-folder old-path data BEFORE any update so targeted cache eviction is possible
     final List<Map<String, Object>> subFolderSnapshot = loadSubFolderSnapshot(oldPath, hostId);
 
-    // Capture the old values before mutation so they can be restored if a subsequent DB
-    // operation fails. @WrapInTransaction rolls back DB changes, but Java object state is
-    // not rolled back automatically — callers that retain a reference after failure would
-    // otherwise observe the new name permanently despite the rename not completing.
-    final String oldFolderName = folder.getName();
-    final String oldAssetName  = ident.getAssetName();
+    // Set the new name on the folder object so getNewFolderRecord() picks it up via
+    // initialFolder.getName(). Restore the original name if a subsequent DB operation fails,
+    // since @WrapInTransaction rolls back DB changes but Java object state is not automatically
+    // restored.
+    folder.setName(newName);
 
-    // Update the folder record in-place.
-    // save() uses upsertFolder keyed on the existing inode — the inode is NOT regenerated.
-    // Because the inode is preserved, the old updateOtherFolderReferences calls (which updated
-    // structure.folder, permission.inode_id, permission_reference.asset_id from oldInode to
-    // newInode) are pure no-ops (SET x = X WHERE x = X) and have been removed.
-    //
-    // save() internally calls APILocator.getIdentifierAPI().find() to get the identifier for
-    // FolderCache eviction. At this point the identifier cache still holds asset_name=oldName, so
-    // save() correctly evicts the old path-keyed cache entry. The identifier cache is evicted on
-    // the next line, AFTER save() reads it, preserving the correct ordering.
+    // Create a new folder record
+    // getNewFolderRecord() calls IdentifierAPI.createNew() which generates the deterministic hash
+    // based on assetType:hostname:parentPath:folderName — changing the URL changes the identity.
+    final Folder newFolder;
     try {
-      folder.setName(newName);
-      save(folder);
-
-      // Evict old identifier URI cache entry BEFORE mutating asset_name: removeFromCacheByIdentifier
-      // looks up the identifier by ID from cache (still holds old URI) and evicts by that key.
-      // After save(ident), the new URI key is stored normally.
-      CacheLocator.getIdentifierCache().removeFromCacheByIdentifier(ident.getId());
-      ident.setAssetName(newName);
-      APILocator.getIdentifierAPI().save(ident);
-
-      // Evict again AFTER the DB identifier row has been updated. This prevents a race where the
-      // identifier can be cached by another thread (e.g. reindex) after the pre-update eviction
-      // but before/while the asset_name mutation becomes visible.
-      CacheLocator.getIdentifierCache().removeFromCacheDirect(ident.getId(), hostId, null);
+      newFolder = getNewFolderRecord(folder, APILocator.systemUser(), parentPath, hostId);
     } catch (final DotDataException | RuntimeException e) {
-      folder.setName(oldFolderName);
-      ident.setAssetName(oldAssetName);
+      folder.setName(ident.getAssetName()); // restore original name
       // A concurrent rename to the same destination path can race past the collision check and
       // hit the DB unique constraint on identifier(parent_path, asset_name, host_inode).
       // Treat this the same as a pre-checked collision: restore state and return false.
@@ -849,31 +828,42 @@ public class FolderFactoryImpl extends FolderFactory {
       throw e;
     }
 
-    // Evict identifier cache for the entire subtree BEFORE the bulk UPDATE so the eviction can
-    // reliably find the old-path-keyed cache entries. removeFromCacheByIdentifier() uses the
-    // cached identifier's current URI (old path) to evict the path-keyed entry; if we waited
-    // until after the UPDATE, those URI-keyed entries might already have been LRU-evicted and
-    // the stale old-path keys would survive indefinitely.
+    // Evict identifier cache for the entire old subtree BEFORE the bulk UPDATE so the eviction
+    // can reliably find the old-path-keyed cache entries.
     clearIdentifierCacheForSubtree(oldPath, hostId);
 
     // Bulk-update every child identifier's parent_path, processing parent folders before children.
-    // The snapshot already contains all sub-folder paths so no extra SELECTs are needed.
+    // The new folder identifier already exists in the DB (created above), so the trigger's
+    // parent-path existence check passes at every depth level as parents are updated first.
     updateChildPaths(oldPath, newPath, hostId, subFolderSnapshot);
 
-    // Evict identifier caches rooted at the *new* parent path after the bulk update.
-    // This closes the window where other threads could have cached descendants with the old
-    // parent_path during the rename transaction.
+    // Evict identifier caches rooted at the new path after the bulk update.
     clearIdentifierCacheForSubtree(newPath, hostId);
 
-    // Targeted folder cache eviction for affected sub-folders (using old path data from snapshot)
+    // Targeted folder cache eviction for affected sub-folders (using old path data from snapshot).
     evictSubFolderCache(subFolderSnapshot, hostId);
 
-    // Nav cache cleanup — evict the renamed folder, its parent, and every sub-folder in the tree
+    // Nav cache cleanup — evict the renamed folder, its parent, and every sub-folder in the tree.
+    // Uses folder.getInode() (old inode) before it is replaced below.
     CacheLocator.getNavToolCache().removeNav(folder.getHostId(), folder.getInode());
     CacheLocator.getNavToolCache().removeNavByPath(hostId, parentPath);
     for (final Map<String, Object> row : subFolderSnapshot) {
       CacheLocator.getNavToolCache().removeNav(hostId, (String) row.get("inode"));
     }
+
+    // Transfer content-type structure and permission references from old inode to new inode.
+    updateOtherFolderReferences(newFolder.getInode(), folder.getInode());
+
+    // Delete old folder. All children have been moved to the new path by updateChildPaths(),
+    // so check_child_assets() finds no remaining children pointing to the old path and the
+    // delete succeeds without a constraint violation.
+    delete(folder);
+
+    // Update the caller's folder reference to the new inode and identifier so that
+    // FolderAPIImpl.refreshContentUnderFolder() targets the correct renamed folder.
+    // folder.getName() was already set to newName above.
+    folder.setInode(newFolder.getInode());
+    folder.setIdentifier(newFolder.getIdentifier());
 
     return true;
   }

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
@@ -641,8 +641,8 @@ public class FolderFactoryImpl extends FolderFactory {
         "UPDATE permission SET inode_id = ? WHERE inode_id = ?", newFolderInode, oldFolderInode);
     dc.executeUpdate(
         "UPDATE permission_reference SET asset_id = ? WHERE asset_id = ?", newFolderInode, oldFolderInode);
-    CacheLocator.getPermissionCache().remove(oldFolderInode);
-    CacheLocator.getPermissionCache().remove(newFolderInode);
+    APILocator.getPermissionAPI().removePermissionableFromCache(oldFolderInode);
+    APILocator.getPermissionAPI().removePermissionableFromCache(newFolderInode);
   }
 
   /**

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
@@ -634,12 +634,15 @@ public class FolderFactoryImpl extends FolderFactory {
 
   private void updateOtherFolderReferences(final String newFolderInode, final String oldFolderInode)
       throws DotDataException {
-    new DotConnect().executeUpdate(
+    final DotConnect dc = new DotConnect();
+    dc.executeUpdate(
         "UPDATE structure SET folder = ? WHERE folder = ?", newFolderInode, oldFolderInode);
-    new DotConnect().executeUpdate(
+    dc.executeUpdate(
         "UPDATE permission SET inode_id = ? WHERE inode_id = ?", newFolderInode, oldFolderInode);
-    new DotConnect().executeUpdate(
+    dc.executeUpdate(
         "UPDATE permission_reference SET asset_id = ? WHERE asset_id = ?", newFolderInode, oldFolderInode);
+    CacheLocator.getPermissionCache().remove(oldFolderInode);
+    CacheLocator.getPermissionCache().remove(newFolderInode);
   }
 
   /**

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
@@ -637,6 +637,9 @@ public class FolderFactoryImpl extends FolderFactory {
     final DotConnect dc = new DotConnect();
     dc.executeUpdate(
         "UPDATE structure SET folder = ? WHERE folder = ?", newFolderInode, oldFolderInode);
+    APILocator.getContentTypeAPI(APILocator.systemUser())
+        .search("folder='" + newFolderInode + "'", "mod_date", -1, 0)
+        .forEach(CacheLocator.getContentTypeCache2()::remove);
     dc.executeUpdate(
         "UPDATE permission SET inode_id = ? WHERE inode_id = ?", newFolderInode, oldFolderInode);
     dc.executeUpdate(

--- a/dotcms-integration/src/test/java/com/dotmarketing/portlets/folders/business/FolderAPITest.java
+++ b/dotcms-integration/src/test/java/com/dotmarketing/portlets/folders/business/FolderAPITest.java
@@ -151,10 +151,11 @@ public class FolderAPITest extends IntegrationTestBase {//24 contentlets
 
 
 	/**
-	 * Verifies the basic in-place rename behavior for a folder hierarchy.
+	 * Verifies the rename behavior for a folder hierarchy.
 	 * <p>
-	 * All folder identifiers are preserved (same UUIDs), while
-	 * {@code asset_name}/{@code parent_path} are updated consistently.
+	 * The renamed folder receives a new identifier (URL change = identity change). Sub-folder
+	 * identifiers are preserved but their {@code parent_path} values are updated consistently
+	 * via bulk SQL — no Elasticsearch dependency, so unindexed items are never missed.
 	 */
 	@Test
 	public void renameFolder() throws Exception {
@@ -167,26 +168,33 @@ public class FolderAPITest extends IntegrationTestBase {//24 contentlets
 		final Folder ftest3 = folderAPI
 				.createFolders(ftest.getPath()+"/ff1/ff2/ff3", host, user, false);
 
+		// Capture the old top-level identifier BEFORE rename — the factory mutates ftest.
+		final String oldFtestIdentifier = ftest.getIdentifier();
+
 		final String newFolderName = "folderTestXX" + System.currentTimeMillis();
 		assertTrue(folderAPI.renameFolder(ftest, newFolderName, user, false));
 
-		final Identifier ident  = identifierAPI.loadFromDb(ftest.getIdentifier());
+		// Old top-level identifier is deleted (create+delete cycle).
+		final Identifier oldIdent = identifierAPI.loadFromDb(oldFtestIdentifier);
+		assertNull(oldIdent);
+
+		// Renamed folder has a new identifier UUID.
+		assertNotEquals(oldFtestIdentifier, ftest.getIdentifier());
+
+		// Sub-folder identifiers are preserved — only parent_path is updated via bulk SQL.
 		final Identifier ident1 = identifierAPI.loadFromDb(ftest1.getIdentifier());
 		final Identifier ident2 = identifierAPI.loadFromDb(ftest2.getIdentifier());
 		final Identifier ident3 = identifierAPI.loadFromDb(ftest3.getIdentifier());
-
-		assertNotNull(ident);
 		assertNotNull(ident1);
 		assertNotNull(ident2);
 		assertNotNull(ident3);
-
-		assertEquals(newFolderName, ident.getAssetName());
 
 		final String newRootPath = StringPool.SLASH + newFolderName + StringPool.SLASH;
 		assertEquals(newRootPath, ident1.getParentPath());
 		assertEquals(newRootPath + "ff1" + StringPool.SLASH, ident2.getParentPath());
 		assertEquals(newRootPath + "ff1" + StringPool.SLASH + "ff2" + StringPool.SLASH, ident3.getParentPath());
 
+		// Folder and sub-folders are findable by new paths.
 		final Folder newFolder = folderAPI.findFolderByPath(newRootPath, host, user, false);
 		final Folder newFolder1 = folderAPI.findFolderByPath(newRootPath + "ff1/", host, user, false);
 		final Folder newFolder2 = folderAPI.findFolderByPath(newRootPath + "ff1/ff2/", host, user, false);
@@ -196,7 +204,10 @@ public class FolderAPITest extends IntegrationTestBase {//24 contentlets
 		assertNotNull(newFolder1);
 		assertNotNull(newFolder2);
 		assertNotNull(newFolder3);
+		// Top-level folder has a new identifier; ftest object was mutated to hold it.
 		assertEquals(ftest.getIdentifier(), newFolder.getIdentifier());
+		assertNotEquals(oldFtestIdentifier, newFolder.getIdentifier());
+		// Sub-folders retain their original identifiers.
 		assertEquals(ftest1.getIdentifier(), newFolder1.getIdentifier());
 		assertEquals(ftest2.getIdentifier(), newFolder2.getIdentifier());
 		assertEquals(ftest3.getIdentifier(), newFolder3.getIdentifier());
@@ -239,7 +250,8 @@ public class FolderAPITest extends IntegrationTestBase {//24 contentlets
 	 *     <li><b>Expected Result:</b>
 	 *     <ul>
 	 *         <li>Rename returns {@code true}.</li>
-	 *         <li>Folder and sub-folder identifiers are unchanged (in-place rename).</li>
+	 *         <li>The renamed folder receives a new identifier UUID (URL change = identity change).</li>
+	 *         <li>Sub-folder identifier is preserved (only {@code parent_path} is updated).</li>
 	 *         <li>All direct children's {@code parent_path} in the DB equals the new folder path.</li>
 	 *         <li>All sub-children's {@code parent_path} equals the new sub-folder path.</li>
 	 *         <li>Folder is findable by new path; old path resolves to nothing.</li>
@@ -268,6 +280,7 @@ public class FolderAPITest extends IntegrationTestBase {//24 contentlets
 		// File asset under /original/sub/
 		final Contentlet fileInSub = new FileAssetDataGen(subFolder, "content-in-sub").nextPersisted();
 
+		// Capture old identifiers BEFORE rename — factory mutates the parentFolder object.
 		final String parentIdentifierId = parentFolder.getIdentifier();
 		final String subIdentifierId    = subFolder.getIdentifier();
 
@@ -277,15 +290,17 @@ public class FolderAPITest extends IntegrationTestBase {//24 contentlets
 
 		assertTrue("renameFolder must return true", renamed);
 
-		// Folder identifier is preserved — in-place rename, no create+delete cycle
-		final Identifier renamedParentIdent = identifierAPI.loadFromDb(parentIdentifierId);
-		assertNotNull("Parent folder identifier must still exist after rename", renamedParentIdent);
-		assertEquals("Parent folder identifier UUID must be unchanged",
-				parentIdentifierId, renamedParentIdent.getId());
-		assertEquals("Parent folder asset_name must reflect the new name",
-				newName, renamedParentIdent.getAssetName());
+		// Renamed folder gets a new identifier — create+delete cycle, URL change = identity change.
+		final Identifier oldParentIdent = identifierAPI.loadFromDb(parentIdentifierId);
+		assertNull("Old parent folder identifier must be deleted after rename", oldParentIdent);
+		assertNotEquals("Renamed folder must have a new identifier UUID",
+				parentIdentifierId, parentFolder.getIdentifier());
+		final Identifier newParentIdent = identifierAPI.loadFromDb(parentFolder.getIdentifier());
+		assertNotNull("New parent folder identifier must exist after rename", newParentIdent);
+		assertEquals("New parent folder asset_name must reflect the new name",
+				newName, newParentIdent.getAssetName());
 
-		// Sub-folder identifier is preserved and its parent_path is updated
+		// Sub-folder identifier is preserved and its parent_path is updated via bulk SQL.
 		final Identifier renamedSubIdent = identifierAPI.loadFromDb(subIdentifierId);
 		assertNotNull("Sub-folder identifier must still exist", renamedSubIdent);
 		assertEquals("Sub-folder identifier UUID must be unchanged",
@@ -311,11 +326,13 @@ public class FolderAPITest extends IntegrationTestBase {//24 contentlets
 		assertEquals("File asset in sub-folder: parent_path must reflect the renamed sub-folder path",
 				"/" + newName + "/sub/", fileInSubIdent.getParentPath());
 
-		// Folder is findable by new path and returns the same identifier
+		// Folder is findable by new path and has the new identifier UUID.
 		final Folder foundByNewPath = folderAPI.findFolderByPath("/" + newName + "/", site, user, false);
 		assertNotNull("Renamed folder must be findable by new path", foundByNewPath);
-		assertEquals("Folder found by new path must have the same identifier UUID",
+		assertNotEquals("Folder found by new path must have a new identifier UUID (URL change = identity change)",
 				parentIdentifierId, foundByNewPath.getIdentifier());
+		assertEquals("Folder found by new path must match the mutated parentFolder identifier",
+				parentFolder.getIdentifier(), foundByNewPath.getIdentifier());
 
 		// Old path no longer resolves to a folder — verified via DB to avoid a stale cache hit.
 		// A DB query for asset_name=originalName under parentPath='/' must return no folder row.

